### PR TITLE
Enable previewing the color scheme in the command palette

### LINKED ISSUE #6689: Allow commands to be previewed in the command palette
Use cases:
* Switching schemes - I want to be able to see what the scheme will look like, _before_ I commit to it
* Changing opacity - I'm thinking we have a nested "Change opacity..."/["opacity: 0%", "opacity: 5%", "opacity: 10%"...] commands, and as they key through them, it updates in real time
* Probably useful with the tab switcher

These would need to be commands that could be _undo-able_. Commands where we could build a "reset" command given the _current_ state of the Terminal, and hitting <kbd>esc</kbd> to dismiss the palette would revert to the original state.

I'd imagine that not all commands could be previewed. Can't preview "open a new tab". 

If you went from one previewable command to another of the same action type, then we don't need to reset in between. But going from a "see theme to foo"->"set opacity to 5%" commands, we'd need to undo the first, then cache the state for the second command.

Could probably have some `ActionArgs` implement `IPreviewable`, with a `ActionAndArgs GenerateUndoCommand(AppLogic);` method. Anything that's `"preview":true` would need to implement that interface.

\</showerthought>

Considered use cases:
* Switching to another color scheme
* changing the font size
* changing the font face
* changing the opacity of acrylic
* Switching tabs (like in the ATS, #6732)

### PR DESCRIPTION
**Pull request by zadjii-msft**
## Summary of the Pull Request

Allow schemes to be previewed as the user hovers over them in the Command Palette.

preview-set-color-scheme

## References
* Branched off of #8392, which is why the commit history is so polluted. 330a8e8 : 544b2fd has the interesting commits
* #5400: cmdpal megathread

### Potential follow-ups
* changing the font size
* changing the font face
* changing the opacity of acrylic

## PR Checklist
* [x] Closes #6689, a last straggling FHL PR
* [x] I work here
* [ ] Tests added/passed
* [n/a] Requires documentation to be updated - I don't think so

## Detailed Description of the Pull Request / Additional comments

This works by inserting a "preview" `TerminalSettings` into the settings hierarchy, before the `TermControl`'s runtime settings, and after the ones from the actual `CascadiaSettings`. This allows us to modify that preview settings object, then discard it when we're done with the preview.

This could also be used for other settings in the future - I built it to be extensible to other `ShortcutAction`s, though I haven't implemented those yet.

## Validation Steps Performed

* Select a colorscheme - it becomes the active one
* `colortool -x <scheme>` after selecting a scheme - colortool overrides the selected scheme
* Select a colorscheme after a `colortool -x <scheme>` after selecting a scheme - the scheme in the palette becomes the active one
* Pressing <kbd>esc</kbd> at any point to dismiss the command palette - scheme returns to the previous one
* reloading the settings - returns to the scheme in the settings

### CODE REVIEW

**Comment by miniksa** [comment]
image

Yikes.

File: src/cascadia/TerminalApp/TerminalPage.cpp
Comment (miniksa) [inline src/cascadia/TerminalApp/TerminalPage.cpp:2894]: GH todo linked?

File: src/cascadia/TerminalApp/TerminalPage.cpp
Comment (miniksa) [inline src/cascadia/TerminalApp/TerminalPage.cpp:2762]: Hmmmm I sort of wanted a throw/failfast on the default case here so we could catch when someone adds another preview action but forgets to clean it up since that's no good.

But the more I read... it looks like preview and end preview are always being called for actions and the ones that don't support it just fall through, yeah?

**Review by miniksa — CHANGES_REQUESTED** [review]

File: src/cascadia/TerminalApp/TerminalPage.cpp
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/TerminalPage.cpp:2762]: Yeppers.

File: src/cascadia/TerminalApp/TerminalPage.cpp
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/TerminalPage.cpp:2894]: Meh, I didn't file one but I'll do it

Comment (zadjii-msft) [reply src/cascadia/TerminalApp/TerminalPage.cpp:2894]: https://github.com/microsoft/terminal/issues/9818

**Review by carlos-zamora — CHANGES_REQUESTED** [review]
How does this interact with the color scheme overrides in `Profile`?

@PankajBhojwani Mind taking a look at the `TerminalSettingsCreateResult` stuff here?

I'm a bit confused about why we're replacing the root `TerminalSettings`. Mind explaining it a bit more here?

File: src/cascadia/TerminalApp/TerminalPage.cpp
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/TerminalPage.cpp:2849]: sure, that's safer

File: src/cascadia/TerminalApp/TerminalPage.cpp
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/TerminalPage.cpp:2843]: Maybe? If there weren't a parent for the control's settings anymore, that feels like a bit of a big change

File: src/cascadia/TerminalApp/TerminalPage.cpp
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/TerminalPage.cpp:2838]: I guess I'm happily relying on the fact that there's only ever at most three layers here, and there's definitely always two.
```
root settings (from CascadiaSettings)
^
|
previewed settings (may or may not exist)
^
|
control's runtime settings
```

We're mostly going to the root here just to make sure that we get the original settings, so that _later_, in EndPreview, we can reset the parent of the control to the original root settings. If there were ever more layers here, this would start to break down, sure.

Comment (zadjii-msft) [reply src/cascadia/TerminalApp/TerminalPage.cpp:2838]: I might be able to write a test for this that would catch if we broke this assumption

**Review by miniksa — APPROVED** [review]
I'm good but I defer to the more-expert WinRT people to review that stuff before merge.

File: src/cascadia/TerminalApp/AppActionHandlers.cpp
Comment (PankajBhojwani) [inline src/cascadia/TerminalApp/AppActionHandlers.cpp:404]: Have a question about this - why do we clear the runtime colors? I understood the 'control-decided' settings to only be mutable by the control. 

Also, wouldn't this mean that the 'PreviewColorScheme' appearance may not match up to the 'SetColorScheme' appearance? Because it seems that we _don't_ clear out the runtime colors when previewing the new scheme, but we do when we set the new scheme (so if there were some colors in the runtime settings, they will show up in the preview but will disappear after setting the new scheme)

**Review by PankajBhojwani — CHANGES_REQUESTED** [review]
Blocking just for the runtime colors question!

Also I think you already know this, [but the color scheme preview will not work if the profile has an `unfocusedAppearance`](https://github.com/microsoft/terminal/issues/9799#issuecomment-819461755). I'm okay with fixing that separately/in-post though since it looks like we need to have some discussion on how we want to tackle that.

File: src/cascadia/TerminalApp/AppActionHandlers.cpp
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/AppActionHandlers.cpp:404]: Yea, this is a rough one. It probably should only be mutable by the control, but if it _is_, then the SetColorScheme action would do _nothing at all_ when the user's already run `colortool`, which is not really what we want, right? So the only option here is to clear out the runtime settings.

Weirdly enough, this seemed to work fine when doing the colortool -> preview a scheme. I agree that it seems like it shouldn't. I could dig into it and investigate further if need be

File: src/cascadia/TerminalApp/TerminalPage.cpp
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/TerminalPage.cpp]: Okay, I added tests, yay!

File: src/cascadia/TerminalApp/TerminalPage.cpp
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/TerminalPage.cpp:2755]:
```suggestion
```
nit

**Review by carlos-zamora — APPROVED** [review]
The tests make me feel a lot more comfortable. Thank you! 😊

**Review by PankajBhojwani — APPROVED** [review]

**Pull request merged**
